### ISSUE TITLE
Container keeps crashing

### ISSUE THREAD

**Issue by T-vK** [cjimti/iotwifi#12]
* **I'm submitting a ...**
  - [x] bug report
  - [ ] feature request
  - [ ] support request

I have a Raspberry Pi 3 Model B and a Raspberry Pi 3 Model B+.
And I flashed `2018-04-18-raspbian-stretch-lite.img` to the Micro SD cards of both of them (using Etcher).
Both Pis should host an AP and one Pi should connect the other Pis AP while the other Pi should connect to my home WiFi. The docker containers keep crashing almost immediately, though.

I documented everything I did:

I booted both of them up (for the first time with the newly flashed images) and used raspi-config to:

- set the keyboard layout to German
- set the timezone to Germany
- set the WiFi country to BO
- change the default password
- change the host name
- enable SSH

then I rebooted and ran:

``` bash
sudo apt update
sudo apt upgrade
sudo apt install vim screen git nmap
```

then for the Pi 3b I changed `/etc/wpa_supplicant/wpa_supplicant.conf` to:

```
ctrl_interface=DIR=/var/run/wpa_supplicant GROUP=netdev
update_config=1
country=BO

network={
    ssid="MyHomeNetwork"
    psk="PasswordOfMyHomeNetwork"
}
```

for the Pi 3b+ I used:

```
ctrl_interface=DIR=/var/run/wpa_supplicant GROUP=netdev
update_config=1
country=BO

network={
    ssid="rpi3bwap"
    psk="iotwifipass"
}
```

Then I ran:

``` bash
sudo systemctl mask wpa_supplicant.service
sudo pkill wpa_supplicant
curl -sSL https://get.docker.com | sh
sudo usermod -aG docker pi
sudo reboot
```

After the reboot I ran 

``` bash
docker run --rm hello-world
```

which worked as expected.  
Then I ran:  

``` bash
docker pull cjimti/iotwifi
```

Then I created `./wificfg.json` and for the Pi 3b I set it to:

```
{
    "dnsmasq_cfg": {
      "address": "/#/192.168.27.1",
      "dhcp_range": "192.168.27.100,192.168.27.150,1h",
      "vendor_class": "set:device,IoT"
    },
    "host_apd_cfg": {
       "ip": "192.168.27.1",
       "ssid": "rpi3bwap",
       "wpa_passphrase":"iotwifipass",
       "channel": "11"
    },
      "wpa_supplicant_cfg": {
        "cfg_file": "/etc/wpa_supplicant/wpa_supplicant.conf"
    }
}
```

For the Pi 3b+ I set it to:

```
{
    "dnsmasq_cfg": {
      "address": "/#/192.168.28.1",
      "dhcp_range": "192.168.28.100,192.168.28.150,1h",
      "vendor_class": "set:device,IoT"
    },
    "host_apd_cfg": {
       "ip": "192.168.28.1",
       "ssid": "rpi3bpwap",
       "wpa_passphrase":"iotwifipass",
       "channel": "11"
    },
      "wpa_supplicant_cfg": {
        "cfg_file": "/etc/wpa_supplicant/wpa_supplicant.conf"
    }
}
```

I used channel 11 just in case, because that is the same channel as my home WiFi. For the Pi 3b+ I used 192.168.28.x instead of 192.168.27.x, also just in case in matters.

Then I ran:

``` bash
docker run -d --restart=unless-stopped --name=wifi --privileged --net host -v $(pwd)/wificfg.json:/cfg/wificfg.json -v /etc/wpa_supplicant/wpa_supplicant.conf:/etc/wpa_supplicant/wpa_supplicant.conf cjimti/iotwifi
```

(Note that I'm passing the wpa_supplicant.conf through.)

Since it didn't work at that point I rebooted and then I ran:

``` bash
docker stop wifi
docker rm wifi
docker run -d --restart=unless-stopped --name=wifi --privileged --net host -v $(pwd)/wificfg.json:/cfg/wificfg.json -v /etc/wpa_supplicant/wpa_supplicant.conf:/etc/wpa_supplicant/wpa_supplicant.conf cjimti/iotwifi
```

But it was still not working correctly on either of the Pis.

On the Pi 3b+ the Docker container keeps crashing. The output of `docker ps -a` was:

``` 
CONTAINER ID        IMAGE               COMMAND             CREATED              STATUS                      PORTS               NAMES                                  
b625404fad52        cjimti/iotwifi      "/wifi"             About a minute ago   Exited (2) 40 seconds ago                       wifi  
```

It crashes after around 30-60 seconds.

Here is the output of `docker logs wifi` on the Pi 3b+: [rpi3bp.log](https://github.com/cjimti/iotwifi/files/2020281/rpi3bp.log) (MAC addresses and SSIDs censored)

(log of the original post: [docker.log](https://github.com/cjimti/iotwifi/files/2019308/docker.log) (MAC addresses and SSIDs censored))

On the Pi 3b it was slightly different. The container keeps running, but it spits out massive amounts of log output and it doesn't actually work...  
Here is the output of `docker logs wifi` on the Pi 3b: [rpi3b.log](https://github.com/cjimti/iotwifi/files/2020283/rpi3b.log) (MAC addresses and SSIDs censored)

**Comment by cjimti** [comment]
@T-vK I'll take a more in-depth look at your logs later tonight or tomorrow when I have more free time. However, in the meantime double check that wpa_supplicant is not running on the host. I saw a report that `systemctl mask wpa_supplicant.service` did not prevent it from starting on some installations.

Another thing to do is set the container to `--restart unless-stopped`. There are conditions where the interface is not yet ready,  in this case, the container crashes on purpose so that it can be restarted container and re-initialize with any network changes.

**Comment by T-vK** [comment]
I just tried it again from scratch and I noticed that the behavior on the Pi 3b and Pi 3b+ differs. It doesn't work on both of them, but maybe for different reasons. I made some edits to the original post adding separate logs for each Pi. Also, I actually did run the commands with `--restart=unless-stopped`, I should have copy pasted it sorry. 

About the wpa_supplicant. I checked on both Pis while the docker container was stopped if it was running:

On the Pi 3b:
```
sudo ps -aux | grep wpa_supp
root       405  0.0  0.4  10140  3880 ?        Ss   10:52   0:03 wpa_supplicant -B -c/etc/wpa_supplicant/wpa_supplicant.conf -iwlan0 -Dnl80211,wext
```

On the Pi 3b+:
```
sudo ps -aux | grep wpa_supp
root       404  0.0  0.3  10128  3360 ?        Ss   13:35   0:00 wpa_supplicant -B -c/etc/wpa_supplicant/wpa_supplicant.conf -iwlan0 -Dnl80211,wext
root       772  0.0  0.2  10028  2116 ?        Ss   13:35   0:00 wpa_supplicant -B -c/etc/wpa_supplicant/wpa_supplicant.conf -iuap0 -Dnl80211,wext
```

On both Pis `sudo systemctl status wpa_supplicant.service` gave the following output:
```
● wpa_supplicant.service
   Loaded: masked (/dev/null; bad)
   Active: inactive (dead)
```


After killing the `wpa_supplicant` proocesses using `sudo killall wpa_supplicant`, I ran 
```
docker rm wifi
docker run -d --restart=unless-stopped --name=wifi --privileged --net host -v $(pwd)/wificfg.json:/cfg/wificfg.json -v /etc/wpa_supplicant/wpa_supplicant.conf:/etc/wpa_supplicant/wpa_supplicant.conf cjimti/iotwifi
```

on both Pis. 

But it didn't seem to change anything. It's still crashing on the Pi 3b+...

One thing I forgot to mention in the first post is that the AP of the Pi 3b does show up, but I am not able to connect to it with any device (I'm not getting incorrect password errors, it must be something else.)...
The AP of the Pi 3b+ does not show up. In fact on the Pi 3b+ `ifconfig` doesn't even list wlan0 or uap0, but I guess that makes sense when the container crashes...

**Comment by T-vK** [comment]
I just tested it on a Pi Zero W V1.1 as well and the result was pretty much the same as with the Pi 3b. (Container keeps running, but gives massive amounts of log output and doesn't actually connect to my WiFi or host an AP.) Here is the log for the Zero W 1.1: 
[rpizw11.log](https://github.com/cjimti/iotwifi/files/2020513/rpizw11.log)

For this Pi I first had to connect to my WiFi the normal non-docker way and then update/install everything and pull the docker image and then disable and kill wpa_supplicant... Then I proceeded creating/running the container with the same command as before.

**Comment by cjimti** [comment]
@T-vK I can confirm that it works on Pi 2, 3b and 3b+ since I have over 800 3b and a few dozen 3b+ Pis running for a client. However, I don't use the n00ds distro so it's possible that something changed since I wrote the tutorial. It looks to me that you set it up correctly but it's hard to know what is going on with the host. I assume it set up the host network correctly? 

Also, If I understand you correctly, after stopping the container your greps show wpa_supplicant running and in the second case running multiples. wpa_supplicant cannot be running on the host. 

I am going to close this ticket but we can keep the conversation going if we can identify a bug that 
I can fix.

**Comment by T-vK** [comment]
I've never heard of the n00ds distro. I am using the Raspbian Stretch Lite image as provided on the official website: https://www.raspberrypi.org/downloads/raspbian/ (I'm using the 2018-04-18 version)

I'm not sure if the host network was set up correctly, but I don't think so. The hosted AP of the Pi 3 does show up when I do a WiFi scan, but I cannot connect to it with any device, it just fails every time (I ensured the password I entered was the correct one btw) and when I enter ifconfig, it lists wlan0 and uap0. 
On the Pi 3b+ and Zero W 1.1 the docker container just crashes and their hosted APs don't show up when I do WiFi scans. Also, when I run ifconfig on the two, it doesn't list wlan0 or uap0 (probably because the docker container crashed.

> Also, If I understand you correctly, after stopping the container your greps show wpa_supplicant running and in the second case running multiples. wpa_supplicant cannot be running on the host.
Correct. I did kill the processes before running the docker containers.

**Comment by T-vK** [comment]
@cjimti Have you looked at the logs yet? Any ideas what could be wrong? Can you test it on one of your devices with Raspbian Stretch Lite?

**Comment by cjimti** [comment]
@T-vK Sorry I did not see anything I'll install Raspbian Stretch Lite on a device and check it out. Unfortunately, my work has been slammed at the moment working late nights and early mornings so I might not be able to do it until this weekend. The only other thing I could say at the moment would be to completely remove wap_supplicant from the device to ensure that there is no way can interfere.

**Comment by cjimti** [comment]
After a quick test on one of my devices using a Raspian image, I noticed that wpa_supplicant is fired up not only on boot (but you removed that) but also when the interface is brought up. I don't know if the same thing is happening to you, but if so, you can simply remove wpa_supplicant from the host altogether. 

I still plan to install Raspian Lite this weekend and take another look.

**Comment by T-vK** [comment]
I just tested it on my Pi 3b+ by renaming the executable using `sudo mv /sbin/wpa_supplicant /sbin/wpa_supplicant.bkp`. But it was still not working.
I did however make an interesting observation: I accidentally misspelled the SSID of my home WiFi so that the Pi couldn't connect. When that happened, the hosted AP of that Pi actually worked. I was able to connect to it with my Android phone and was able to ssh into it.
When I corrected the SSID and then stopped+removed the docker container and created+ran it again, it went back to old broken state.

**Comment by T-vK** [comment]
Hey @cjimti have you found some time to take another look at it?

**Comment by cjimti** [comment]
@T-vK Unfortunately I am 100% committed to a work project for the next month. I did, however, download Raspian Lite and ran it on a Pi 3 b+. I noticed that masking wap_supplicant is definitely not sufficient since upping the wlan0 interface causes it to start on the host. I see however that you renamed the file, and that should work. After removing wpa_supplicant from the host I was having trouble connecting to hostapd if it was not on channel 6, I don't know if that is just because other networks around me are crowded or some bug, but I was not able to look into it further. Otherwise, I was able to connect, issue a curl call with the credentials for one the wifi networks and it connected. 

There is a lot of room for improvement and added stability, however, I am not experiencing complete failure like you are. 

I am looking for some people interested in working on a version 2, but that is delayed a month at this point.
